[mlir][gpu][spirv] Remove rotation semantics of gpu.shuffle up/down

Hsiangkai opened the pull request:
From the description of gpu.shuffle operation, shuffle up/down rotates values in the subgroup because it applies modulo on the shifted value to calculate the result lane ID. It is inconsistent with the definition of SPIR-V shuffle up/down and NVVM data movement definitions within subgroup.

In NVVM, it says

"If the computed source lane index j is in range, the returned i32 value will be the value of %a from lane j; otherwise, it will be the the value of %a from the current thread."

It will keep the original value if the result land ID is out of range.

In SPIR-V OpGroupNonUniformShuffleUp and OpGroupNonUniformShuffleDown, it says

"The resulting value is undefined if Delta is greater than the current invocation’s id within the scope or if the identified invocation is not in scope restricted tangle."

It's an undefined value if the result land ID is out of range.

Anyway, there is no circular movement in shuffle up/down from these 2 specifications. This patch removes the circular movement in gpu.shuffle up/down and lower gpu.shuffle up/down to SPIR-V OpGroupNonUniformShuffleUp and OpGroupNonUniformShuffleDown directly.

Reference:
https://docs.nvidia.com/cuda/archive/12.2.1/nvvm-ir-spec/index.html#data-movement
https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformShuffleUp
https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformShuffleDown

IgWod-IMG commented:
> In addition, update the spirv.GroupNonUniformRotateKHR assembly format to be consistent with other gpu non-uniform operations.

Apologies for being pedantic, but it'd probably be better to submit this as a separate PR. It feels a bit out of place to submit it with this change. But I'm still relatively new to the project, so I'm not sure what are the (written and unwritten) rules around it :)

Hsiangkai committed:
[mlir][gpu][spirv] Add patterns for gpu.shuffle up/down

Convert

gpu.shuffle down %val, %offset, %width

to

spirv.GroupNonUniformRotateKHR <Subgroup> %val, %offset, cluster_size(%width)

Convert

gpu.shuffle up %val, %offset, %width

to

%down_offset = arith.subi %width, %offset
spirv.GroupNonUniformRotateKHR <Subgroup> %val, %down_offset, cluster_size(%width)

Hsiangkai commented:
> > In addition, update the spirv.GroupNonUniformRotateKHR assembly format to be consistent with other gpu non-uniform operations.
> 
> Apologies for being pedantic, but it'd probably be better to submit this as a separate PR. It feels a bit out of place to submit it with this change. But I'm still relatively new to the project, so I'm not sure what are the (written and unwritten) rules around it :)

Make sense. I created another PR for reformatting. https://github.com/llvm/llvm-project/pull/139224

Hsiangkai committed:
The width argument cannot exceed the subgroup limit.

Hsiangkai committed:
fix typo

Hsiangkai committed:
remove test for gpu.shuffle width != subgroup_size limit

Hsiangkai commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -430,10 +430,12 @@ LogicalResult GPUShuffleConversion::matchAndRewrite(
   IntegerAttr widthAttr;
+  // The width argument specifies the number of lanes that participate in the
+  // shuffle. The width value should not exceed the subgroup limit.
   if (!matchPattern(shuffleOp.getWidth(), m_Constant(&widthAttr)) ||
-      widthAttr.getValue().getZExtValue() != subgroupSize)
+      widthAttr.getValue().getZExtValue() > subgroupSize)
     return rewriter.notifyMatchFailure(
-        shuffleOp, "shuffle width and target subgroup size mismatch");
+        shuffleOp, "shuffle width is larger than target subgroup size");
```
If the width exceeds the subgroup limit, it's a failure case. That's the condition checking.

Hsiangkai replied:
Sorry, I didn't notice that comment. Thanks for pointing it out. For spirv.OpGroupNonUniformShuffle and spirv.OpGroupNonUniformShuffleXor, it makes sense to restrict the width to be the same as subgroup size. However, there is an optional argument `ClusterSize` in spirv.OpGroupNonUniformRotateKHR and `ClusterSize` is not necessary to be equal to the subgroup size.

You can refer the specification of spirv.OpGroupNonUniformRotateKHR
https://github.khronos.org/SPIRV-Registry/extensions/KHR/SPV_KHR_subgroup_rotate.html

```
RotationGroupSize = ClusterSize when ClusterSize is present, otherwise
RotationGroupSize = SubgroupMaxSize if the Kernel capability is declared and SubgroupSize if not.
```
and
```
If ClusterSize is greater than the declared SubGroupSize, executing this instruction results in undefined behavior.
```

I will update the patch according to the above description.

Hsiangkai replied:
After revisiting spirv and nvvm specification, we should remove the rotation semantic in gpu.shuffle up/down.
nvvm:
https://docs.nvidia.com/cuda/archive/12.2.1/nvvm-ir-spec/index.html#data-movement
spirv:
https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformShuffleUp
https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformShuffleDown

In addition, we need another gpu.rotate operation for rotation semantics. I will have another patch for rotation operation.

Hsiangkai committed:
Remove rotation semantic in gpu.shufflw up/down

There is no such semantic in SPIRV OpGroupNonUniformShuffleUp and
OpGroupNonUniformShuffleDown. In addition, there is no such semantic in
NVVM shfl intrinsics.

Refer to NVVM IR spec
https://docs.nvidia.com/cuda/archive/12.2.1/nvvm-ir-spec/index.html#data-movement

"If the computed source lane index j is in range, the returned i32 value
will be the value of %a from lane j; otherwise, it will be the the value
of %a from the current thread. If the thread corresponding to lane j is
inactive, then the returned i32 value is undefined."

Hsiangkai commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
````diff
@@ -1341,7 +1342,8 @@ def GPU_ShuffleOp : GPU_Op<
     ```
 
-    For lane `k`, returns the value from lane `(k - 1) % width`.
+    For lane `k`, returns the value from lane `(k - cst1)`. The resulting value
````
Reply above.

Hsiangkai commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
```diff
@@ -1332,7 +1332,8 @@ def GPU_ShuffleOp : GPU_Op<
 
-    For lane `k`, returns the value from lane `(k + 1) % width`.
+    For lane `k`, returns the value from lane `(k + cst1)`. The resulting value
+    is undefined if the lane is out of bounds in the subgroup.
```
Reply above.

Hsiangkai commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
````diff
@@ -1332,7 +1332,8 @@ def GPU_ShuffleOp : GPU_Op<
     ```
 
-    For lane `k`, returns the value from lane `(k + 1) % width`.
+    For lane `k`, returns the value from lane `(k + cst1)`. The resulting value
````
I updated the title and commit message.

kuhar commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
````diff
@@ -1332,7 +1332,8 @@ def GPU_ShuffleOp : GPU_Op<
     ```
 
-    For lane `k`, returns the value from lane `(k + 1) % width`.
+    For lane `k`, returns the value from lane `(k + cst1)`. The resulting value
````
Could we say that the oob values are poison? We want to avoid using handwavey terms like `unspecified`/`undefined` that don't have precise meaning. For example, what happens when you try to perform arithmetic or branch on them?

Hsiangkai replied:
Updated.

Hsiangkai committed:
Refine description and set 'valid' flag according to the resulting landID

Hsiangkai committed:
refactor and update wording

Hsiangkai commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -416,6 +416,30 @@ LogicalResult GPUBarrierConversion::matchAndRewrite(
+}
+
+Value getLaneId(OpBuilder &rewriter, MLIRContext *ctx, Location loc) {
+  Value dimX = getDimOp<gpu::BlockDimOp>(rewriter, ctx, loc, gpu::Dimension::x);
+  Value dimY = getDimOp<gpu::BlockDimOp>(rewriter, ctx, loc, gpu::Dimension::y);
+  Value tidX = getDimOp<gpu::ThreadIdOp>(rewriter, ctx, loc, gpu::Dimension::x);
+  Value tidY = getDimOp<gpu::ThreadIdOp>(rewriter, ctx, loc, gpu::Dimension::y);
+  Value tidZ = getDimOp<gpu::ThreadIdOp>(rewriter, ctx, loc, gpu::Dimension::z);
+  auto i32Type = rewriter.getIntegerType(32);
+  Value tmp1 = rewriter.create<arith::MulIOp>(loc, i32Type, tidZ, dimY);
+  Value tmp2 = rewriter.create<arith::AddIOp>(loc, i32Type, tmp1, tidY);
+  Value tmp3 = rewriter.create<arith::MulIOp>(loc, i32Type, tmp2, dimX);
+  Value laneId = rewriter.create<arith::AddIOp>(loc, i32Type, tmp3, tidX);
```
According to SPIR-V specification,

```The set of invocations partitioned in some execution models (e.g. GLCompute, Kernel) as a workgroup. Its size is defined statically by either the WorkgroupSize [built-in](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#BuiltIn)```

We should be able to calculate invocation ID from workgroup sizes and thread IDs under GLCompute, and Kernel execution mode. By the way, there is no conversion pattern to convert gpu::LaneId to SPIR-V. I will add a pattern for it first.

Hsiangkai replied:
I just found gpu::LaneIdOp conversion pattern landed recently. I will rebase my patch and leverage gpu::LaneIdOp in this patch.

kuhar replied:
> In addition, MLIR converts `gpu::BlockDimOp` to SPIR-V workgroup size for x, y, z dimensions and converts `gpu::ThreadIdOp` to SPIR-V local invocation Id along x, y, z dimensions.

you can convert thread id to lane id if you know the workgroup size, but the math used to be incorrect

kuhar replied:
How have you tested this code? I wonder if we should add some integration tests since this is may not be trivial to get right.

Hsiangkai committed:
Use gpu::LaneIdOp

Muzammiluddin-Syed-ECE commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -436,25 +436,53 @@ LogicalResult GPUShuffleConversion::matchAndRewrite(
+        loc, scope, adaptor.getValue(), adaptor.getOffset());
+
+    Value laneId = rewriter.create<gpu::LaneIdOp>(loc, widthAttr);
+    Value resultLandId =
```
nit: typo

Hsiangkai replied:
Fixed.

Muzammiluddin-Syed-ECE commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -436,25 +436,53 @@ LogicalResult GPUShuffleConversion::matchAndRewrite(
+        loc, scope, adaptor.getValue(), adaptor.getOffset());
+
+    Value laneId = rewriter.create<gpu::LaneIdOp>(loc, widthAttr);
+    Value resultLandId =
```
nit: typo

Hsiangkai replied:
Fixed

Muzammiluddin-Syed-ECE commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -436,25 +436,53 @@ LogicalResult GPUShuffleConversion::matchAndRewrite(
+
+    Value laneId = rewriter.create<gpu::LaneIdOp>(loc, widthAttr);
+    Value resultLandId =
+        rewriter.create<arith::SubIOp>(loc, laneId, adaptor.getOffset());
```
optional: could add another check to ensure that `adaptor.getOffset()` doesn't have a negative value but that's also done in the verifier for `verifyGroupNonUniformShuffleOp`

Hsiangkai replied:
Added

Hsiangkai committed:
Fix typo and add signless/unsigned checking for offset

Hsiangkai committed:
Use assert instead of if checking

Hsiangkai commented on mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:
```diff
@@ -435,26 +435,59 @@ LogicalResult GPUShuffleConversion::matchAndRewrite(
+  // Ensure the offset is a signless/unsigned integer.
+  if (adaptor.getOffset().getType().isSignedInteger())
+    return rewriter.notifyMatchFailure(
+        shuffleOp, "shuffle offset must be a signless/unsigned integer");
```
Done.

Hsiangkai commented:
Any other comments? Can I merge this patch?
